### ISSUE TITLE
How to finetune on custom dataset

### ISSUE THREAD

**Issue by hrsmanian** [jzhang38/TinyLlama#48]
How does on finetune the model on custom dataset? Should we modify finetune.py?

**Comment by jzhang38** [comment]
TinyLlama is compatible with Llama1/2. You can hassle-free fine-tune our model on many FT codebases built for llama, such as: 
1. https://github.com/lm-sys/FastChat
2. https://github.com/imoneoi/openchat
3. https://github.com/artidoro/qlora

If you use a different data format, you might need to modify some code.

**Comment by VatsaDev** [comment]
Do you plan to release a Qlora based finetune that could work with something simple, like a text file?

**Comment by RonanKMcGovern** [comment]
@VatsaDev can't you just use transformers bitsandbytes and do QLoRA like any other Llama model?

**Comment by VatsaDev** [comment]
@RonanKMcGovern I tried running it off the Qlora script for oasst guanaco on the repo, but whenever I go for a custom dataset, its tends to fail

**Comment by RonanKMcGovern** [comment]
ok, I had no trouble, maybe try the bitsandbytes script: https://colab.research.google.com/drive/1VoYNfYDKcKRQRor98Zbf2-9VQTtGJ24k?usp=sharing

**Comment by VatsaDev** [comment]
@RonanKMcGovern, thanks! try it out soon

**Comment by VatsaDev** [comment]
@RonanKMcGovern  The script takes 135 hours to tokenize the dataset? Its around 20gb, seems unnaturally long

**Comment by RonanKMcGovern** [comment]
What dataset are you using.

Sorry, I meant Tim Dettmers filtered openassistanr

**Comment by VatsaDev** [comment]
https://huggingface.co/datasets/VatsaDev/UnagamiData, a personal dataset, its around ~13m rows of chat sentences, kinda long, it takes two hours to prepare with my custom data loader. For HF datasets implementation with NanoGPT, It takes 8 hours, but with the bnb script, it always goes to around 135hrs

**Comment by RonanKMcGovern** [comment]
Hmm @VatsaDev I'm not entirely sure but I do see this message:
```
You're using a GPTNeoXTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.
```
So there may be something about the default configuration that is slowing things down owing to padding. Perhaps your custom method also does a better job of tokenizing in batches (that's a wild guess though, I believe that batching doesn't necessarily improve tokenization speed a lot). Sorry I'm not of more use here.

**Comment by VatsaDev** [comment]
Thanks for the effort. 

The custom method uses Tiktoken, which is supposably faster, check on that error msg.
